cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/realtime/source/SqsSource.java

```diff
@@ -0,0 +1,175 @@
+import javax.jms.TextMessage;
+
+/**
+ * Realtime source that reads from Amazon SQS.
+ */
+@Plugin(type = "source")
+@Name("Sqs")
+@Description("Sqs Realtime Source - Emits a record with a field 'body' of string type")
+public class SqsSource extends RealtimeSource<StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(SqsSource.class);
+  private static final String REGION_DESCRIPTION = "Region where the queue is located.";
+  private static final String ACCESSKEY_DESCRIPTION = "Access Key of the AWS account to use.";
+  private static final String ACCESSID_DESCRIPTION = "Access ID of the AWS account to use.";
+  private static final String QUEUENAME_DESCRIPTION = "Name of the queue.";
+  private static final String ENDPOINT_DESCRIPTION = "Endpoint of the SQS server to connect to. Leave blank to " +
+    "connect to AWS.";
+  private static final Schema DEFAULT_SCHEMA = Schema.recordOf(
+    "event",
+    Schema.Field.of("body", Schema.of(Schema.Type.STRING))
+  );
+  private static final int MAX_MESSSAGE_COUNT = 20;
+  private static final int TIMEOUT_LENGTH = 1000;
+  private final SqsConfig config;
+  private SQSConnectionFactory connectionFactory;
+  private MessageConsumer consumer;
+  private Session session;
+  private SQSConnection connection;
+  public SqsSource(SqsConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void initialize(RealtimeContext contex) {
+    try {
+      SQSConnectionFactory.Builder sqsBuild = SQSConnectionFactory.builder()
+        .withRegion(Region.getRegion(Regions.fromName(config.region)));
+      connectionFactory = config.endpoint == null ? sqsBuild.build() : sqsBuild.withEndpoint(config.endpoint).build();
+      connection = connectionFactory.createConnection(config.accessID, config.accessKey);
+      session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+      consumer = session.createConsumer(session.createQueue(config.queueName));
+      connection.start();
+    } catch (Exception e) {
+      if (session != null) {
+        try {
+          session.close();
+        } catch (Exception ex1) {
+          LOG.warn("Exception when closing session", ex1);
+        }
+      }
+      if (connection != null) {
+        try {
+          connection.close();
+        } catch (Exception ex2) {
+          LOG.warn("Exception when closing connection", ex2);
+        }
+      }
+      if (consumer != null) {
+        try {
+          consumer.close();
+        } catch (Exception ex3) {
+          LOG.warn("Exception when closing consumer", ex3);
+        }
+      }
+      LOG.error("Failed to connect to SQS");
+      throw new IllegalStateException("Could not connect to SQS.");
+    }
+  }
+
+  @Override
+  public SourceState poll(Emitter<StructuredRecord> writer, SourceState currentState) throws Exception {
+    Message msg;
+    int count = 0;
+    while ((msg = consumer.receive(TIMEOUT_LENGTH)) != null && count < MAX_MESSSAGE_COUNT) {
```

Comment:
Sorry, meant empty as in null message, not empty string.  When you do it this way, the client will only select a subset of the servers and check if any messages are on those servers.  Long polling will check all servers of a queue for some time, and return all of them.  We might want to keep SQS separate, because it is unique in this way.  Thoughts?